[BUG] `jax.value_and_grad` fails when the differentiated circuit has a complex matrix under `QubitUnitary`

```python
dev = qml.device("lightning.qubit", wires=1)

@qml.qjit
@qml.qnode(dev)
def circuit(x):
	qml.QubitUnitary(jnp.array([[x,1j],[-1,0]]), wires=[0])
	return qml.probs()

def cost(data):
	return circuit(data)[0]

print("value_and_grad: ", 
	catalyst.value_and_grad(cost)(1.1)
	)

>>>
Traceback (most recent call last):
  File "/home/paul.wang/small_playgrounds_dump/symmetry_demo/unimplemented_array_format.py", line 29, in <module>
    catalyst.value_and_grad(cost)(1.1)
  File "/home/paul.wang/catalyst/frontend/catalyst/api_extensions/differentiation.py", line 623, in __call__
    results = jax.value_and_grad(self.fn, argnums=argnums)(*args)
  File "/home/paul.wang/small_playgrounds_dump/symmetry_demo/unimplemented_array_format.py", line 24, in cost
    return circuit(data)[0]
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 511, in __call__
    return self.jaxed_function(*args, **kwargs)
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 833, in __call__
    return self.jaxed_function(*args, **kwargs)
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 809, in compute_jvp
    derivatives = self.wrap_callback(self.get_derivative_qjit(argnums), *primals)
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 789, in get_derivative_qjit
    self.derivative_functions[argnum_key] = QJIT(
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 498, in __init__
    self.aot_compile()
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 530, in aot_compile
    self.jaxpr, self.out_treedef, self.c_sig = self.capture(self.user_sig or ())
  File "/home/paul.wang/catalyst/frontend/catalyst/debug/instruments.py", line 143, in wrapper
    return fn(*args, **kwargs)
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 628, in capture
    jaxpr, treedef = trace_to_jaxpr(
  File "/home/paul.wang/catalyst/frontend/catalyst/jax_tracer.py", line 353, in trace_to_jaxpr
    jaxpr, out_treedef = make_jaxpr2(func, **make_jaxpr_kwargs)(*args, **kwargs)
  File "/home/paul.wang/catalyst/frontend/catalyst/jax_extras/tracing.py", line 402, in make_jaxpr_f
    jaxpr, output_type, consts = trace_to_jaxpr_dynamic2(f)
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 783, in deriv_wrapper
    return catalyst.jacobian(self.qjit_function, argnum=argnums)(*args, **kwargs)
  File "/home/paul.wang/catalyst/frontend/catalyst/api_extensions/differentiation.py", line 613, in __call__
    results = grad_p.bind(*args_data, jaxpr=jaxpr, fn=fn, grad_params=grad_params)
jax._src.source_info_util.JaxStackTraceBeforeTransformation: ValueError: unimplemented array format conversion from format: Zd

The preceding stack trace is the source of the JAX operation that, once transformed by JAX, triggered the following exception.

--------------------

The above exception was the direct cause of the following exception:

jax.errors.SimplifiedTraceback: For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/paul.wang/small_playgrounds_dump/symmetry_demo/unimplemented_array_format.py", line 29, in <module>
    catalyst.value_and_grad(cost)(1.1)
  File "/home/paul.wang/catalyst/frontend/catalyst/api_extensions/differentiation.py", line 623, in __call__
    results = jax.value_and_grad(self.fn, argnums=argnums)(*args)
  File "/home/paul.wang/small_playgrounds_dump/symmetry_demo/unimplemented_array_format.py", line 24, in cost
    return circuit(data)[0]
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 511, in __call__
    return self.jaxed_function(*args, **kwargs)
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 833, in __call__
    return self.jaxed_function(*args, **kwargs)
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 809, in compute_jvp
    derivatives = self.wrap_callback(self.get_derivative_qjit(argnums), *primals)
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 789, in get_derivative_qjit
    self.derivative_functions[argnum_key] = QJIT(
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 498, in __init__
    self.aot_compile()
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 533, in aot_compile
    self.mlir_module, self.mlir = self.generate_ir()
  File "/home/paul.wang/catalyst/frontend/catalyst/debug/instruments.py", line 143, in wrapper
    return fn(*args, **kwargs)
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 642, in generate_ir
    mlir_module, ctx = lower_jaxpr_to_mlir(self.jaxpr, self.__name__)
  File "/home/paul.wang/catalyst/frontend/catalyst/jax_tracer.py", line 381, in lower_jaxpr_to_mlir
    mlir_module, ctx = jaxpr_to_mlir(func_name, jaxpr)
  File "/home/paul.wang/catalyst/frontend/catalyst/jax_extras/lowering.py", line 65, in jaxpr_to_mlir
    module, context = custom_lower_jaxpr_to_module(
  File "/home/paul.wang/catalyst/frontend/catalyst/jax_extras/lowering.py", line 133, in custom_lower_jaxpr_to_module
    lower_jaxpr_to_fun(
  File "/home/paul.wang/catalyst/frontend/catalyst/jax_primitives.py", line 411, in _grad_lowering
    constants = [
  File "/home/paul.wang/catalyst/frontend/catalyst/jax_primitives.py", line 412, in <listcomp>
    StableHLOConstantOp(ir.DenseElementsAttr.get(np.asarray(const))).results
ValueError: unimplemented array format conversion from format: Zd
```

The same error persists when using `jax.value_and_grad` (expected, since the `catalyst.value_and_grad` is called outside a `@qjit`).

Posiibly, this is due to mlir not having support for complex arrays. 

-------------------------------------------------------------------------------------------
When switching to a real matrix, this error disappears:
```python
dev = qml.device("lightning.qubit", wires=1)

@qml.qjit
@qml.qnode(dev)
def circuit(x):
	qml.QubitUnitary(jnp.array([[x,1],[-1,0]]), wires=[0])
	return qml.probs()

def cost(data):
	return circuit(data)[0]

print("value_and_grad: ", 
	catalyst.value_and_grad(cost)(1.1)
	)

>>>
value_and_grad:  (Array(1.21, dtype=float64), Array(0., dtype=float64))
```

On a circuit that is not qjit-compiled, with a complex matrix, this error also disappears:
```python
#@qml.qjit
@qml.qnode(dev)
def circuit(x):
	qml.QubitUnitary(jnp.array([[x,1j],[-1,0]]), wires=[0])
	return qml.probs()

def cost(data):
	return circuit(data)[0]

print(jax.value_and_grad(cost)(1.1))
>>>(Array(3.74939946e-33, dtype=float64), Array(nan, dtype=float64, weak_type=True))
```

> @paul0403 does this bug remain when `value_and_grad` is _inside_ the qjit? Typically we aim to do gradients within the qjit if we can:
> 
> ```python
> dev = qml.device("lightning.qubit", wires=1)
> 
> 
> @qml.qnode(dev)
> def circuit(x):
>     qml.QubitUnitary(jnp.array([[x,1],[-1,0]]), wires=[0])
>     return qml.probs()
> 
> @qml.qjit
> def f(data):
>     cost = lambda x: circuit(x)[0]
>     return catalyst.value_and_grad(cost)(data)
> 
> print("value_and_grad: ", f(1.1))
> ```

This particular snippet bugs with 
```
Traceback (most recent call last):
  File "/home/paul.wang/catalyst/hi.py", line 20, in <module>
    print("value_and_grad: ", f(1.1))
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 505, in __call__
    requires_promotion = self.jit_compile(args)
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 572, in jit_compile
    self.jaxpr, self.out_treedef, self.c_sig = self.capture(args)
  File "/home/paul.wang/catalyst/frontend/catalyst/debug/instruments.py", line 143, in wrapper
    return fn(*args, **kwargs)
  File "/home/paul.wang/catalyst/frontend/catalyst/jit.py", line 628, in capture
    jaxpr, treedef = trace_to_jaxpr(
  File "/home/paul.wang/catalyst/frontend/catalyst/jax_tracer.py", line 353, in trace_to_jaxpr
    jaxpr, out_treedef = make_jaxpr2(func, **make_jaxpr_kwargs)(*args, **kwargs)
  File "/home/paul.wang/catalyst/frontend/catalyst/jax_extras/tracing.py", line 402, in make_jaxpr_f
    jaxpr, output_type, consts = trace_to_jaxpr_dynamic2(f)
  File "/home/paul.wang/.local/lib/python3.10/site-packages/jax/_src/profiler.py", line 336, in wrapper
    return func(*args, **kwargs)
  File "/home/paul.wang/.local/lib/python3.10/site-packages/jax/_src/interpreters/partial_eval.py", line 2324, in trace_to_jaxpr_dynamic2
    jaxpr, out_type, consts = trace_to_subjaxpr_dynamic2(fun, main, debug_info)
  File "/home/paul.wang/.local/lib/python3.10/site-packages/jax/_src/interpreters/partial_eval.py", line 2339, in trace_to_subjaxpr_dynamic2
    ans = fun.call_wrapped(*in_tracers_)
  File "/home/paul.wang/.local/lib/python3.10/site-packages/jax/_src/linear_util.py", line 191, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "/home/paul.wang/catalyst/hi.py", line 18, in f
    return catalyst.value_and_grad(cost)(data)
  File "/home/paul.wang/catalyst/frontend/catalyst/api_extensions/differentiation.py", line 592, in __call__
    assert (
AssertionError: Tracing of value_and_grad is not implemented yet
```

If we switch to just `grad` instead of `value_and_grad`, then a more mysterious bug appears (even when the matrix is real instead of complex), with no python traceback and just an IR dump:
```
error: <unknown>:0:0: in function preprocess_<lambda>.cloned void (ptr, ptr, i64, i64, i64, ptr, ptr, i64, ptr, ptr, i64): Enzyme: Cannot deduce type of memset   call void @llvm.memset.p0.i64(ptr noundef nonnull align 8 dereferenceable(16) %.fca.2.gep6, i8 0, i64 16, i1 false) #6
<analysis>
  %.fca.2.gep6 = getelementptr inbounds { ptr, ptr, i64, [1 x i64], [1 x i64] }, ptr %15, i64 0, i32 2: {[-1]:Pointer}, intvals: {}
ptr %0: {[-1]:Pointer}, intvals: {}
ptr %1: {[-1]:Pointer}, intvals: {}
i64 %2: {[-1]:Integer}, intvals: {}
i64 %3: {[-1]:Integer}, intvals: {}
i64 %4: {[-1]:Integer}, intvals: {}
ptr %5: {[-1]:Pointer}, intvals: {}
ptr %6: {[-1]:Pointer}, intvals: {}
i64 %7: {[-1]:Integer}, intvals: {}
ptr %8: {[-1]:Pointer}, intvals: {}
ptr %9: {[-1]:Pointer}, intvals: {}
i64 %10: {[-1]:Integer}, intvals: {}
  %.fca.1.gep = getelementptr inbounds { ptr, ptr, i64, [1 x i64], [1 x i64] }, ptr %17, i64 0, i32 1: {[-1]:Pointer, [-1,0]:Pointer}, intvals: {}
  %17 = alloca { ptr, ptr, i64, [1 x i64], [1 x i64] }, align 8: {[-1]:Pointer, [-1,0]:Pointer, [-1,8]:Pointer, [-1,24]:Integer, [-1,25]:Integer, [-1,26]:Integer, [-1,27]:Integer, [-1,28]:Integer, [-1,29]:Integer, [-1,30]:Integer, [-1,31]:Integer, [-1,32]:Integer, [-1,33]:Integer, [-1,34]:Integer, [-1,35]:Integer, [-1,36]:Integer, [-1,37]:Integer, [-1,38]:Integer, [-1,39]:Integer}, intvals: {}
  %14 = alloca { ptr, ptr, i64 }, align 8: {[-1]:Pointer, [-1,0]:Pointer, [-1,8]:Pointer, [-1,16]:Integer, [-1,17]:Integer, [-1,18]:Integer, [-1,19]:Integer, [-1,20]:Integer, [-1,21]:Integer, [-1,22]:Integer, [-1,23]:Integer}, intvals: {}
i64 0: {[-1]:Anything}, intvals: {0,}
  %15 = alloca { ptr, ptr, i64, [1 x i64], [1 x i64] }, align 8: {[-1]:Pointer, [-1,0]:Pointer, [-1,8]:Pointer, [-1,32]:Integer, [-1,33]:Integer, [-1,34]:Integer, [-1,35]:Integer, [-1,36]:Integer, [-1,37]:Integer, [-1,38]:Integer, [-1,39]:Integer}, intvals: {}
  %18 = load double, ptr %16, align 8: {}, intvals: {}
  call void @circuit.quantum(ptr nonnull %13, ptr nonnull %14, ptr nonnull %15, ptr nonnull %17) #6: {}, intvals: {}
  %.fca.2.gep = getelementptr inbounds { ptr, ptr, i64, [1 x i64], [1 x i64] }, ptr %17, i64 0, i32 2: {[-1]:Pointer}, intvals: {}
  %.fca.3.0.gep24 = getelementptr inbounds { ptr, ptr, i64, [1 x i64], [1 x i64] }, ptr %13, i64 0, i32 3, i64 0: {[-1]:Pointer, [-1,0]:Integer, [-1,1]:Integer, [-1,2]:Integer, [-1,3]:Integer, [-1,4]:Integer, [-1,5]:Integer, [-1,6]:Integer, [-1,7]:Integer}, intvals: {}
  %.fca.4.0.gep26 = getelementptr inbounds { ptr, ptr, i64, [1 x i64], [1 x i64] }, ptr %13, i64 0, i32 4, i64 0: {[-1]:Pointer, [-1,0]:Integer, [-1,1]:Integer, [-1,2]:Integer, [-1,3]:Integer, [-1,4]:Integer, [-1,5]:Integer, [-1,6]:Integer, [-1,7]:Integer}, intvals: {}
  %.fca.4.0.gep10 = getelementptr inbounds { ptr, ptr, i64, [1 x i64], [1 x i64] }, ptr %15, i64 0, i32 4, i64 0: {[-1]:Pointer, [-1,0]:Integer, [-1,1]:Integer, [-1,2]:Integer, [-1,3]:Integer, [-1,4]:Integer, [-1,5]:Integer, [-1,6]:Integer, [-1,7]:Integer}, intvals: {}
  %16 = tail call ptr @_mlir_memref_to_llvm_alloc(i64 16) #6: {[-1]:Pointer}, intvals: {}
  %.fca.1.gep20 = getelementptr inbounds { ptr, ptr, i64, [1 x i64], [1 x i64] }, ptr %13, i64 0, i32 1: {[-1]:Pointer, [-1,0]:Pointer}, intvals: {}
  %.fca.1.gep14 = getelementptr inbounds { ptr, ptr, i64 }, ptr %14, i64 0, i32 1: {[-1]:Pointer, [-1,0]:Pointer}, intvals: {}
  %13 = alloca { ptr, ptr, i64, [1 x i64], [1 x i64] }, align 8: {[-1]:Pointer, [-1,0]:Pointer, [-1,8]:Pointer, [-1,16]:Integer, [-1,17]:Integer, [-1,18]:Integer, [-1,19]:Integer, [-1,20]:Integer, [-1,21]:Integer, [-1,22]:Integer, [-1,23]:Integer, [-1,24]:Integer, [-1,25]:Integer, [-1,26]:Integer, [-1,27]:Integer, [-1,28]:Integer, [-1,29]:Integer, [-1,30]:Integer, [-1,31]:Integer, [-1,32]:Integer, [-1,33]:Integer, [-1,34]:Integer, [-1,35]:Integer, [-1,36]:Integer, [-1,37]:Integer, [-1,38]:Integer, [-1,39]:Integer}, intvals: {}
  %.fca.2.gep16 = getelementptr inbounds { ptr, ptr, i64 }, ptr %14, i64 0, i32 2: {[-1]:Pointer, [-1,0]:Integer, [-1,1]:Integer, [-1,2]:Integer, [-1,3]:Integer, [-1,4]:Integer, [-1,5]:Integer, [-1,6]:Integer, [-1,7]:Integer}, intvals: {}
  %12 = tail call ptr @_mlir_memref_to_llvm_alloc(i64 0) #6: {[-1]:Pointer}, intvals: {}
i64 2: {[-1]:Integer}, intvals: {2,}
  %.fca.4.0.gep = getelementptr inbounds { ptr, ptr, i64, [1 x i64], [1 x i64] }, ptr %17, i64 0, i32 4, i64 0: {[-1]:Pointer, [-1,0]:Integer, [-1,1]:Integer, [-1,2]:Integer, [-1,3]:Integer, [-1,4]:Integer, [-1,5]:Integer, [-1,6]:Integer, [-1,7]:Integer}, intvals: {}
  %.fca.3.0.gep = getelementptr inbounds { ptr, ptr, i64, [1 x i64], [1 x i64] }, ptr %17, i64 0, i32 3, i64 0: {[-1]:Pointer, [-1,0]:Integer, [-1,1]:Integer, [-1,2]:Integer, [-1,3]:Integer, [-1,4]:Integer, [-1,5]:Integer, [-1,6]:Integer, [-1,7]:Integer}, intvals: {}
i64 1: {[-1]:Integer}, intvals: {1,}
  %.fca.1.gep4 = getelementptr inbounds { ptr, ptr, i64, [1 x i64], [1 x i64] }, ptr %15, i64 0, i32 1: {[-1]:Pointer, [-1,0]:Pointer}, intvals: {}
  %.fca.2.gep22 = getelementptr inbounds { ptr, ptr, i64, [1 x i64], [1 x i64] }, ptr %13, i64 0, i32 2: {[-1]:Pointer, [-1,0]:Integer, [-1,1]:Integer, [-1,2]:Integer, [-1,3]:Integer, [-1,4]:Integer, [-1,5]:Integer, [-1,6]:Integer, [-1,7]:Integer}, intvals: {}
</analysis>

```

--- a/doc/changelog.md
+++ b/doc/changelog.md
@@ -137,6 +137,9 @@
   In this bug fix, four openblas routines were newly linked and are now discoverable by `stablehlo.custom_call@<blas_routine>`. They are `blas_dtrsm`, `blas_ztrsm`, `lapack_dgetrf`, `lapack_zgetrf`.
   [(#752)](https://github.com/PennyLaneAI/catalyst/pull/752)
 
+* Correctly recording types of constant array when lowering `catalyst.grad` to mlir
+  [(#778)](https://github.com/PennyLaneAI/catalyst/pull/778)
+
 <h3>Internal changes</h3>
 
 * Add support to use a locally cloned PennyLane Lightning repository with the runtime.

--- a/frontend/catalyst/jax_primitives.py
+++ b/frontend/catalyst/jax_primitives.py
@@ -413,11 +413,10 @@ def _grad_lowering(ctx, *args, jaxpr, fn, grad_params):
     constants = []
     for const in jaxpr.consts:
         const_type = shape_dtype_to_ir_type(const.shape, const.dtype)
-        constants.append(
-            StableHLOConstantOp(
-                ir.DenseElementsAttr.get(np.asarray(const), type=const_type)
-            ).results
-        )
+        nparray = np.asarray(const)
+        attr = ir.DenseElementsAttr.get(nparray, type=const_type)
+        constantVals = StableHLOConstantOp(attr).results
+        constants.append(constantVals)
     args_and_consts = constants + list(args)
 
     return GradOp(

--- a/frontend/catalyst/jax_primitives.py
+++ b/frontend/catalyst/jax_primitives.py
@@ -457,14 +457,10 @@ def _jvp_lowering(ctx, *args, jaxpr, fn, grad_params):
 
     output_types = list(map(mlir.aval_to_ir_types, ctx.avals_out))
     flat_output_types = util.flatten(output_types)
-    constants = []
-    for const in jaxpr.consts:
-        const_type = shape_dtype_to_ir_type(const.shape, const.dtype)
-        constants.append(
-            StableHLOConstantOp(
-                ir.DenseElementsAttr.get(np.asarray(const), type=const_type)
-            ).results
-        )
+    constants = [
+        StableHLOConstantOp(ir.DenseElementsAttr.get(np.asarray(const))).results
+        for const in jaxpr.consts
+    ]
     consts_and_args = constants + args
     func_call_jaxpr = jaxpr.eqns[0].params["call_jaxpr"]
     func_args = consts_and_args[: len(func_call_jaxpr.invars)]
@@ -517,14 +513,10 @@ def _vjp_lowering(ctx, *args, jaxpr, fn, grad_params):
 
     output_types = list(map(mlir.aval_to_ir_types, ctx.avals_out))
     flat_output_types = util.flatten(output_types)
-    constants = []
-    for const in jaxpr.consts:
-        const_type = shape_dtype_to_ir_type(const.shape, const.dtype)
-        constants.append(
-            StableHLOConstantOp(
-                ir.DenseElementsAttr.get(np.asarray(const), type=const_type)
-            ).results
-        )
+    constants = [
+        StableHLOConstantOp(ir.DenseElementsAttr.get(np.asarray(const))).results
+        for const in jaxpr.consts
+    ]
     consts_and_args = constants + args
     func_call_jaxpr = jaxpr.eqns[0].params["call_jaxpr"]
     func_args = consts_and_args[: len(func_call_jaxpr.invars)]

--- a/frontend/catalyst/jax_primitives.py
+++ b/frontend/catalyst/jax_primitives.py
@@ -28,6 +28,7 @@ from jax._src.lib.mlir import ir
 from jax.core import AbstractValue
 from jax.interpreters import mlir
 from jax.tree_util import PyTreeDef, tree_unflatten
+from jaxlib.hlo_helpers import shape_dtype_to_ir_type
 from jaxlib.mlir.dialects.arith import (
     AddIOp,
     CeilDivSIOp,
@@ -408,10 +409,15 @@ def _grad_lowering(ctx, *args, jaxpr, fn, grad_params):
     # ``ir.DenseElementsAttr.get()`` constructs a dense elements attribute from an array of
     # element values. This doesn't support ``jaxlib.xla_extension.Array``, so we have to cast
     # such constants to numpy array types.
-    constants = [
-        StableHLOConstantOp(ir.DenseElementsAttr.get(np.asarray(const))).results
-        for const in jaxpr.consts
-    ]
+
+    constants = []
+    for const in jaxpr.consts:
+        const_type = shape_dtype_to_ir_type(const.shape, const.dtype)
+        constants.append(
+            StableHLOConstantOp(
+                ir.DenseElementsAttr.get(np.asarray(const), type=const_type)
+            ).results
+        )
     args_and_consts = constants + list(args)
 
     return GradOp(
@@ -451,10 +457,14 @@ def _jvp_lowering(ctx, *args, jaxpr, fn, grad_params):
 
     output_types = list(map(mlir.aval_to_ir_types, ctx.avals_out))
     flat_output_types = util.flatten(output_types)
-    constants = [
-        StableHLOConstantOp(ir.DenseElementsAttr.get(np.asarray(const))).results
-        for const in jaxpr.consts
-    ]
+    constants = []
+    for const in jaxpr.consts:
+        const_type = shape_dtype_to_ir_type(const.shape, const.dtype)
+        constants.append(
+            StableHLOConstantOp(
+                ir.DenseElementsAttr.get(np.asarray(const), type=const_type)
+            ).results
+        )
     consts_and_args = constants + args
     func_call_jaxpr = jaxpr.eqns[0].params["call_jaxpr"]
     func_args = consts_and_args[: len(func_call_jaxpr.invars)]
@@ -507,10 +517,14 @@ def _vjp_lowering(ctx, *args, jaxpr, fn, grad_params):
 
     output_types = list(map(mlir.aval_to_ir_types, ctx.avals_out))
     flat_output_types = util.flatten(output_types)
-    constants = [
-        StableHLOConstantOp(ir.DenseElementsAttr.get(np.asarray(const))).results
-        for const in jaxpr.consts
-    ]
+    constants = []
+    for const in jaxpr.consts:
+        const_type = shape_dtype_to_ir_type(const.shape, const.dtype)
+        constants.append(
+            StableHLOConstantOp(
+                ir.DenseElementsAttr.get(np.asarray(const), type=const_type)
+            ).results
+        )
     consts_and_args = constants + args
     func_call_jaxpr = jaxpr.eqns[0].params["call_jaxpr"]
     func_args = consts_and_args[: len(func_call_jaxpr.invars)]

--- a/frontend/test/pytest/test_gradient.py
+++ b/frontend/test/pytest/test_gradient.py
@@ -1157,7 +1157,7 @@ def test_adj_qubitunitary(inp, backend):
 
     def f(x):
         qml.RX(x, wires=0)
-        U1 = (np.array([[0.5+0.5j, -0.5-0.5j], [0.5-0.5j, 0.5-0.5j]], dtype=complex))
+        U1 = np.array([[0.5 + 0.5j, -0.5 - 0.5j], [0.5 - 0.5j, 0.5 - 0.5j]], dtype=complex)
         qml.QubitUnitary(U1, wires=0)
         return qml.expval(qml.PauliY(0))
 

--- a/frontend/test/pytest/test_gradient.py
+++ b/frontend/test/pytest/test_gradient.py
@@ -1157,7 +1157,11 @@ def test_adj_qubitunitary(inp, backend):
 
     def f(x):
         qml.RX(x, wires=0)
-        U1 = 1 / np.sqrt(2) * np.array([[1.0+2.0j, 3.0-4.5j], [9.8-7.6j, -1.0+3j]], dtype=complex)
+        U1 = (
+            1
+            / np.sqrt(2)
+            * np.array([[1.0 + 2.0j, 3.0 - 4.5j], [9.8 - 7.6j, -1.0 + 3j]], dtype=complex)
+        )
         qml.QubitUnitary(U1, wires=0)
         return qml.expval(qml.PauliY(0))
 

--- a/frontend/test/pytest/test_gradient.py
+++ b/frontend/test/pytest/test_gradient.py
@@ -1157,11 +1157,7 @@ def test_adj_qubitunitary(inp, backend):
 
     def f(x):
         qml.RX(x, wires=0)
-        U1 = (
-            1
-            / np.sqrt(2)
-            * np.array([[1.0 + 2.0j, 3.0 - 4.5j], [9.8 - 7.6j, -1.0 + 3j]], dtype=complex)
-        )
+        U1 = (np.array([[0.5+0.5j, -0.5-0.5j], [0.5-0.5j, 0.5-0.5j]], dtype=complex))
         qml.QubitUnitary(U1, wires=0)
         return qml.expval(qml.PauliY(0))
 
@@ -1173,7 +1169,7 @@ def test_adj_qubitunitary(inp, backend):
 
     def interpreted(x):
         device = qml.device("default.qubit", wires=1)
-        g = qml.QNode(f, device, diff_method="adjoint")
+        g = qml.QNode(f, device, diff_method="backprop")
         h = qml.grad(g, argnum=0)
         return h(x)
 

--- a/frontend/test/pytest/test_gradient.py
+++ b/frontend/test/pytest/test_gradient.py
@@ -1163,13 +1163,13 @@ def test_adj_qubitunitary(inp, backend):
 
     @qjit()
     def compiled(x: float):
-        g = qml.qnode(qml.device(backend, wires=1))(f)
+        g = qml.qnode(qml.device(backend, wires=1), diff_method="adjoint")(f)
         h = grad(g, method="auto")
         return h(x)
 
     def interpreted(x):
         device = qml.device("default.qubit", wires=1)
-        g = qml.QNode(f, device)
+        g = qml.QNode(f, device, diff_method="adjoint")
         h = qml.grad(g, argnum=0)
         return h(x)
 

--- a/frontend/test/pytest/test_gradient.py
+++ b/frontend/test/pytest/test_gradient.py
@@ -1151,26 +1151,25 @@ def test_pytrees_args_return_classical():
     assert np.allclose(flatten_res_jax, flatten_res_catalyst)
 
 
-@pytest.mark.xfail(reason="QubitUnitrary is not support with catalyst.grad")
 @pytest.mark.parametrize("inp", [(1.0), (2.0), (3.0), (4.0)])
 def test_adj_qubitunitary(inp, backend):
     """Test the adjoint method."""
 
     def f(x):
         qml.RX(x, wires=0)
-        U1 = 1 / np.sqrt(2) * np.array([[1.0, 1.0], [1.0, -1.0]], dtype=complex)
+        U1 = 1 / np.sqrt(2) * np.array([[1.0+2.0j, 3.0-4.5j], [9.8-7.6j, -1.0+3j]], dtype=complex)
         qml.QubitUnitary(U1, wires=0)
         return qml.expval(qml.PauliY(0))
 
     @qjit()
     def compiled(x: float):
-        g = qml.qnode(qml.device(backend, wires=1), diff_method="adjoint")(f)
+        g = qml.qnode(qml.device(backend, wires=1))(f)
         h = grad(g, method="auto")
         return h(x)
 
     def interpreted(x):
         device = qml.device("default.qubit", wires=1)
-        g = qml.QNode(f, device, diff_method="backprop")
+        g = qml.QNode(f, device)
         h = qml.grad(g, argnum=0)
         return h(x)
 

